File: backend/src/main/java/com/suplr/backend/service/StorageService.java
package com.suplr.backend.service;

import jakarta.annotation.PostConstruct;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.core.sync.RequestBody;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.s3.S3Client;
import software.amazon.awssdk.services.s3.model.*;

import java.net.URI;

@Slf4j
@Service
public class StorageService {

    @Value("${app.r2.account-id:}")
    private String accountId;

    @Value("${app.r2.access-key-id:}")
    private String accessKeyId;

    @Value("${app.r2.secret-access-key:}")
    private String secretAccessKey;

    @Value("${app.r2.bucket:}")
    private String bucket;

    @Value("${app.r2.public-url:}")
    private String publicUrlBase;

    private S3Client s3Client;

    @PostConstruct
    public void init() {
        if (isNotConfigured()) {
            log.warn("Cloudflare R2 storage credentials are not fully configured.");
            return;
        }

        this.s3Client = S3Client.builder()
                .endpointOverride(URI.create(
                        "https://" + accountId + ".r2.cloudflarestorage.com"))
                .credentialsProvider(StaticCredentialsProvider.create(
                        AwsBasicCredentials.create(accessKeyId, secretAccessKey)))
                .region(Region.of("auto"))
                .build();
    }

    public boolean isNotConfigured() {
        return isBlank(accountId)
                || isBlank(accessKeyId)
                || isBlank(secretAccessKey)
                || isBlank(bucket);
    }

    private boolean isBlank(String str) {
        return str == null || str.isBlank();
    }

    public String upload(String key, byte[] data, String contentType) {
        if (isNotConfigured()) {
            throw new IllegalStateException("R2 storage is not configured.");
        }
        try {
            s3Client.putObject(
                    PutObjectRequest.builder()
                            .bucket(bucket)
                            .key(key)
                            .contentType(contentType)
                            .build(),
                    RequestBody.fromBytes(data)
            );
            log.debug("Uploaded {} ({} bytes) to R2", key, data.length);
            return publicUrl(key);
        } catch (Exception e) {
            log.error("R2 upload failed for key {}: {}", key, e.getMessage());
            throw new RuntimeException("R2 upload failed: " + e.getMessage(), e);
        }
    }

    public boolean exists(String key) {
        if (isNotConfigured()) return false;
        try {
            s3Client.headObject(
                    HeadObjectRequest.builder()
                            .bucket(bucket)
                            .key(key)
                            .build()
            );
            return true;
        } catch (NoSuchKeyException e) {
            return false;
        } catch (Exception e) {
            log.warn("R2 exists check failed for key {}: {}", key, e.getMessage());
            return false;
        }
    }

    public void pingBucket() {
        if (isNotConfigured()) {
            throw new IllegalStateException("R2 not configured");
        }
        s3Client.headBucket(HeadBucketRequest.builder().bucket(bucket).build());
    }

    public String publicUrl(String key) {
        if (isBlank(publicUrlBase)) {
            return key;
        }
        return publicUrlBase.replaceAll("/+$", "") + "/" + key;
    }
}

File: backend/src/storage/storage.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import {
  HeadBucketCommand,
  HeadObjectCommand,
  NoSuchKey,
  PutObjectCommand,
  S3Client,
} from '@aws-sdk/client-s3';

@Injectable()
export class StorageService {
  private readonly logger = new Logger(StorageService.name);
  private s3: S3Client | null = null;
  private readonly bucket: string;
  private readonly publicUrlBase: string;

  constructor(private readonly config: ConfigService) {
    const accountId = config.get<string>('R2_ACCOUNT_ID', '');
    const accessKeyId = config.get<string>('R2_ACCESS_KEY_ID', '');
    const secretAccessKey = config.get<string>('R2_SECRET_ACCESS_KEY', '');
    this.bucket = config.get<string>('R2_BUCKET', '');
    this.publicUrlBase = config.get<string>('R2_PUBLIC_URL', '');

    if (accountId && accessKeyId && secretAccessKey && this.bucket) {
      this.s3 = new S3Client({
        endpoint: `https://${accountId}.r2.cloudflarestorage.com`,
        credentials: { accessKeyId, secretAccessKey },
        region: 'auto',
      });
    } else {
      this.logger.warn('R2 storage not configured');
    }
  }

  isNotConfigured(): boolean {
    return !this.s3;
  }

  upload(key: string, data: Buffer, contentType: string): string {
    if (!this.s3) throw new Error('R2 not configured');
    this.s3
      .send(
        new PutObjectCommand({
          Bucket: this.bucket,
          Key: key,
          Body: data,
          ContentType: contentType,
        }),
      )
      .then(() => this.logger.debug(`Uploaded ${key} (${data.length} bytes)`))
      .catch((e: Error) =>
        this.logger.error(`R2 upload failed for ${key}: ${e.message}`),
      );
    return this.publicUrl(key);
  }

  async exists(key: string): Promise<boolean> {
    if (!this.s3) return false;
    try {
      await this.s3.send(
        new HeadObjectCommand({ Bucket: this.bucket, Key: key }),
      );
      return true;
    } catch (e) {
      if (e instanceof NoSuchKey) return false;
      this.logger.warn(
        `R2 exists check failed for ${key}: ${(e as Error).message}`,
      );
      return false;
    }
  }

  async pingBucket(): Promise<void> {
    if (!this.s3) throw new Error('R2 not configured');
    await this.s3.send(new HeadBucketCommand({ Bucket: this.bucket }));
  }

  publicUrl(key: string): string {
    if (!this.publicUrlBase) return key;
    return `${this.publicUrlBase.replace(/\/+$/, '')}/${key}`;
  }
}
